### PR TITLE
feat: add login throttling to login endpoint

### PR DESCRIPTION
**Pull request by ObodoakorDavid**
## What does the PR do?

This PR is about adding throttling to the login endpoint

How should this be manually tested?
Send a POST request to /api/v1/auth/login repeatedly (4 times to be specific) with a request body as below

```
{
"email": "String",
"password": "String",
}
```

Error Response: 

```
{
    "message": "This User is unable to login now, kindly try again in an hour",
    "error": "Forbidden",
    "status_code": 403
}
```

#### Checklist of what you did

- Implemented throttling on the user login endpoint
- Reset attempt left to the normal amount upon successfully login
- Added unit and e2e tests for user login service and auth service

Reference Issue
[FEAT](https://github.com/hngprojects/hng_boilerplate_nestjs/issues/184) Login Throttling

### CODE REVIEW

**Comment by davidolanrewaju** [comment]
Include what error responses and success responses are gotten (when the endpoint is executed) in your PR

File: src/modules/auth/auth.service.ts [hunk 1]
```diff
@@ -103,10 +104,26 @@ export default class AuthenticationService {
...
+          const oneHourAgo = new Date(Date.now() + 60 * 60 * 1000);
+          if (user.time_left && new Date(user.time_left) > oneHourAgo) {
+            throw new CustomHttpException({ message: USER_BANNED, error: 'Forbidden' }, HttpStatus.FORBIDDEN);
+          }
+
+          user.time_left = new Date();
+          user.attempts_left = 3;
+        }
+
+        await this.userService.updateUserAttempts(user.id, user.attempts_left, user.time_left);
```
Comment (davidolanrewaju) [inline src/modules/auth/auth.service.ts:119]: Wrap your db calls in a try catch block to check for easydebugging

File: src/modules/auth/auth.service.ts
Comment (Heba-WebDev) [inline src/modules/auth/auth.service.ts:1]: I know your issue is basically another step within logging in the user but please don't add this logic in login method. Create a helper method that check the number of intents. Refactor please

File: src/modules/auth/auth.service.ts (hunk 1 above)
Comment (ObodoakorDavid) [reply src/modules/auth/auth.service.ts:122]: It is inside a try and catch.

File: src/modules/auth/auth.service.ts
Comment (ObodoakorDavid) [reply src/modules/auth/auth.service.ts:1]: Done just that. Please have a look and see if this what you mean.

File: src/modules/auth/auth.service.ts [hunk 2]
```diff
@@ -139,6 +142,40 @@ export default class AuthenticationService {
     }
   }
 
+  async handleLoginAttempts(userId: string, isValidAttempt: Boolean) {
+    const user = await this.userService.getUserRecord({
+      identifier: userId,
+      identifierType: 'id',
+    });
+
+    console.log(isValidAttempt);
```
Comment (davidolanrewaju) [inline src/modules/auth/auth.service.ts:151]: Remove all console.log

Comment (davidolanrewaju) [inline src/modules/auth/auth.service.ts:151]: Remove all console.log

Comment (ObodoakorDavid) [reply src/modules/auth/auth.service.ts]: Done that.

File: src/modules/auth/dto/login-user.dto.ts
[the remark this answers is not in the document]
Comment (ObodoakorDavid) [reply src/modules/auth/dto/login-user.dto.ts:1]: Alright. On it.

**Review by Heba-WebDev — APPROVED** [review]

**Pull request closed without merging**